generator/src/test/java/com/scottlogic/deg/generator/utils/RepeatableIteratorTests.java

```diff
@@ -0,0 +1,151 @@
+package com.scottlogic.deg.generator.utils;
+
+import org.junit.Assert;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+class RepeatableIteratorTests {
+    @Test
+    void next_underlyingIteratorHasValueAndFirstTimeCalling_returnsExpectedValue() {
+        RepeatableIterator<String> iterator = getIterator(new ArrayList<String>() {{ add("Test"); }});
+
+        String value = iterator.next();
+
+        Assert.assertEquals("Test", value);
+    }
+
+    @Test
+    void next_underlyingIteratorHasMultipleValuesAndFirstTimeCalling_returnsAllExpectedValues() {
+        RepeatableIterator<String> iterator = getIterator(
+            new ArrayList<String>() {{
+                add("First String");
+                add("Second String");
+                add("Third String");
+            }}
+        );
+
+        String firstValue = iterator.next();
+        String secondValue = iterator.next();
+        String thirdValue = iterator.next();
+
+        Assert.assertEquals(firstValue, "First String");
+        Assert.assertEquals(secondValue, "Second String");
+        Assert.assertEquals(thirdValue, "Third String");
+    }
+
+    @Test
+    void next_underlyingIteratorHasMultipleValuesAndResetCacheCalledAndNextCalled_returnsExpectedValuesMoreThanOnce() {
+        RepeatableIterator<String> iterator = getIterator(
+            new ArrayList<String>() {{
+                add("First String");
+                add("Second String");
+                add("Third String");
+            }}
+        );
+
+        iterator.next();
+        iterator.next();
+        iterator.next();
+        iterator.resetCache();
+        String firstValue = iterator.next();
+        String secondValue = iterator.next();
+        String thirdValue = iterator.next();
+
+        Assert.assertEquals(firstValue, "First String");
+        Assert.assertEquals(secondValue, "Second String");
+        Assert.assertEquals(thirdValue, "Third String");
+    }
+
+    @Test
+    void next_underlyingIteratorHasMultipleValuesAndResetCalledBeforeTheEndOfIterator_returnsAllExpectedValues() {
+        RepeatableIterator<String> iterator = getIterator(
+            new ArrayList<String>() {{
+                add("First String");
+                add("Second String");
+                add("Third String");
+                add("Fourth String");
+            }}
+        );
+
+        iterator.next();
+        iterator.next();
+        iterator.resetCache();
+        String firstValue = iterator.next();
+        String secondValue = iterator.next();
+        String thirdValue = iterator.next();
+        String fourthValue = iterator.next();
+
+        Assert.assertEquals(firstValue, "First String");
+        Assert.assertEquals(secondValue, "Second String");
+        Assert.assertEquals(thirdValue, "Third String");
+        Assert.assertEquals(secondValue, "Second String");
```

Comment:
You don't need this assertion?

Reply:
You're right that you could just call the next methods and if it doesn't throw an exception the test passes. However, it's checking that you get the correct values in the expected order

Reply:
Its more about the fact that you're asserting `secondValue` twice.

Reply:
Oh, completely missed that!